mixer/pkg/checkcache/keyShape.go

```diff
@@ -0,0 +1,239 @@
+			return false
+		}
+
+		// since absentAttrs is sorted by name, continue processing stringMaps until a new name is found.
+		for {
+			_, ok := sm[ks.absentAttrs[i].mapKey]
+			if ok {
+				// if the map key is present, then this bag won't work
+				return false
+			}
+
+			// break loop if at the end or name changes.
+			i++
+			if i == len(ks.absentAttrs) || ks.absentAttrs[i].name != name {
+				break
+			}
+		}
+	}
+
+	// compatible
+	return true
+}
+
+// checkPresentAttrs ensures that the input bag includes all of the attributes that must be present
+func (ks keyShape) checkPresentAttrs(attrs attribute.Bag) bool {
+	for i := 0; i < len(ks.presentAttrs); i++ {
+		name := ks.presentAttrs[i].name
+
+		v, ok := attrs.Get(name)
+		if !ok {
+			// attribute not present, not compatible
+			return false
+		}
+
+		sm, ok := v.(map[string]string)
+		if !ok {
+			// not a string map, so we're good
+			continue
+		}
+
+		// since presentAttrs is sorted by name, continue processing stringMaps until a new name is found.
+		for {
+			_, ok := sm[ks.presentAttrs[i].mapKey]
+			if !ok {
+				// string map key is missing, not compatible
+				return false
+			}
+
+			// break loop if at the end or name changes.
+			i++
+			if i == len(ks.presentAttrs) || ks.presentAttrs[i].name != name {
+				break
+			}
+		}
+	}
+
+	// compatible
+	return true
+}
+
+// makeKey creates a cache key for the given attribute bag.
+// This function assumes that the bag has previously been deemed compatible via isCompatible.
+func (ks keyShape) makeKey(attrs attribute.Bag) string {
+	buf := pool.GetBuffer()
+	b := make([]byte, 8, 8)
+
+	for i := 0; i < len(ks.presentAttrs); i++ {
+		name := ks.presentAttrs[i].name
+
+		buf.WriteString(name)
+		buf.WriteByte(delimiter)
+
+		v, _ := attrs.Get(name) // no need to check for errors, since isCompatible already checked this
+		switch v := v.(type) {
+		case string:
+			buf.WriteString(v)
+
+		case []byte:
+			buf.Write(v)
+
+		case time.Duration:
+		case int64:
+			binary.LittleEndian.PutUint64(b, uint64(v))
+			buf.Write(b)
+
+		case float64:
+			binary.LittleEndian.PutUint64(b, math.Float64bits(v))
+			buf.Write(b)
+
+		case bool:
+			if v {
+				buf.WriteByte(1)
+			} else {
+				buf.WriteByte(0)
+			}
+
+		case time.Time:
+			binary.LittleEndian.PutUint64(b, uint64(v.UnixNano()))
+			buf.Write(b)
+
+		case map[string]string:
+			// Since presentAttrs is sorted by name, continue processing stringMaps until a new name is found.
+			for {
+				v2 := v[ks.presentAttrs[i].mapKey]
+
+				buf.WriteString(ks.presentAttrs[i].mapKey)
+				buf.WriteByte(delimiter)
+				buf.WriteString(v2)
+				buf.WriteByte(delimiter)
+
+				// break loop if at the end or name changes.
+				i++
+				if i == len(ks.presentAttrs) || ks.presentAttrs[i].name != name {
+					break
+				}
+			}
+		}
+
+		buf.WriteByte(delimiter)
+	}
+
+	result := buf.String()
+	pool.PutBuffer(buf)
+
+	return result
```

Comment:
so the key is all concated data,   not sha of the concated data.  Then the key could be very big

Reply:
Yes, the keys can be largish, but avoiding hashing these keys with md5 saves a fair bit of time. I'm going to compress the keys though by leveraging global dictionary indices for most attribute names instead of encoding them as strings (not done yet).